### PR TITLE
Add facades to support juju resolved --all command

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

LOC changes look bigger because the resolved command is moved from cmd/juju/commands to cmd/juju/application since we are using a new facade.

Add a new API method to the Application facade to take new juju resolved params - we can pass in one or more units or --all. This PR supports resolving the supplied units but not --all yet.

$ juju resolved --no-retry mysql/0 
$ juju resolved --no-retry mysql/0 mysql/1

The CLI command will revert to the older client facade method if the new one is not available.
The CLI tests are converted from full stack to unit tests.

## QA steps

Run juju resolved against an old and new server.

## Documentation changes

When the next PR lands we can document the enhanced resolved CLI.

## Bug reference

https://bugs.launchpad.net/bugs/1755141

### CODE REVIEW

File: api/application/client.go [hunk 1]
```diff
@@ -772,3 +772,33 @@ func (c *Client) UnsetApplicationConfig(application string, options []string) er
 	}
 	return results.OneError()
 }
+
+// ResolveUnitErrors clears errors on one or more units.
+// Either specify one or more units, or all.
+func (c *Client) ResolveUnitErrors(units []string, all, retry bool) error {
+	// convert unit name to a tag
+	if len(units) > 0 && all == true {
+		return errors.New("provide units or '--all' but not both")
```
Comment (howbazaar) [inline api/application/client.go]: The API doesn't know about flags. It shouldn't be saying '--all'.

File: api/application/client.go [hunk 2]
```diff
@@ -772,3 +772,33 @@ func (c *Client) UnsetApplicationConfig(application string, options []string) er
...
+	}
+	args := params.UnitsResolved{
+		All:   all,
+		Retry: retry,
+	}
+	if !all {
+		entities := make([]params.Entity, len(units))
```
Comment (howbazaar) [inline api/application/client.go]: Should we be checking for duplicates? The same unit specified twice?

File: cmd/juju/application/resolved.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		Args:    "[<unit> ...]",
+		Purpose: "Marks unit errors resolved and re-executes failed hooks.",
+		Aliases: []string{"resolve"},
+	}
+}
+
+func (c *resolvedCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	f.BoolVar(&c.NoRetry, "no-retry", false, "Do not re-execute failed hooks on the unit")
+	f.BoolVar(&c.All, "all", false, "Marks all units in error as resolved")
```
Comment (howbazaar) [inline cmd/juju/application/resolved.go]: I don't think we should land this branch as it is. We either support --all in the apiserver first, or we don't expose the --all flag on this command until server side is done.

File: cmd/juju/application/resolved.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func (c *resolvedCommand) Init(args []string) error {
+	if c.All {
```
Comment (howbazaar) [inline cmd/juju/application/resolved.go]: I think we could leave this logic here, and just not expose the flag initially.

File: cmd/juju/application/resolved.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+func (c *resolvedCommand) Run(_ *cmd.Context) error {
+	applicationResolveAPI, err := c.getapplicationResolveAPI()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer applicationResolveAPI.Close()
+
+	if applicationResolveAPI.BestAPIVersion() < 6 {
+		if c.All || len(c.UnitNames) > 1 {
+			return errors.Errorf("resolving all units or more than one unit not support by this version of Juju")
```
Comment (howbazaar) [inline cmd/juju/application/resolved.go]: You could just disallow --all, but iterate over the unit names.

File: cmd/juju/application/resolved_test.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		err := s.runResolved(c, t.args)
+		if t.err != "" {
+			c.Assert(err, gc.ErrorMatches, t.err)
+			continue
+		} else {
+			c.Assert(err, jc.ErrorIsNil)
+		}
+		if t.legacyUnit != "" && !t.all {
+			s.mockAPI.CheckCallNames(c, "BestAPIVersion", "Resolved", "Close", "Close")
```
Comment (howbazaar) [inline cmd/juju/application/resolved_test.go]: Why two closes?

Comment (wallyworld) [reply cmd/juju/application/resolved_test.go]: One for Client facade and one for Application facade

File: cmd/juju/application/resolved.go (hunk 5 above)
Comment (wallyworld) [reply cmd/juju/application/resolved.go]: yep, All=false by default so many unit names supported

File: cmd/juju/application/resolved.go (hunk 4 above)
Comment (wallyworld) [reply cmd/juju/application/resolved.go]: yep,  done

File: cmd/juju/application/resolved.go (hunk 3 above)
Comment (wallyworld) [reply cmd/juju/application/resolved.go]: I'll hide the --all on the CLI

File: api/application/client.go (hunk 2 above)
Comment (wallyworld) [reply api/application/client.go]: a check won't hurt, done

File: api/application/client.go (hunk 1 above)
Comment (wallyworld) [reply api/application/client.go]: cut and paste error from original branch, fixed

File: cmd/juju/application/resolved.go (hunk 5 above)
Comment (howbazaar) [reply cmd/juju/application/resolved.go]: I think you misunderstood me. If --all is not allowed, you should still be able to resolve more than one unit by calling the api call once each.

File: cmd/juju/application/resolved_test.go (hunk 6 above)
Comment (howbazaar) [reply cmd/juju/application/resolved_test.go]: Seems a bit strange, and is a side effect of our mocks. Should be documented.

Comment (wallyworld) [reply cmd/juju/application/resolved_test.go]: We open both facades (in this case). When each facade is opened, we do a defer Close(). The mocks are accurately recording the expected api calls here.

File: cmd/juju/application/resolved.go (hunk 5 above)
Comment (wallyworld) [reply cmd/juju/application/resolved.go]: Oh, right ok. Done.

**Comment by wallyworld** [comment]
!!build!!

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**